network-libp2p/src/connection_pool/behaviour.rs

```diff
@@ -113,25 +120,63 @@ struct ConnectionState<T> {
     desired_connections: usize,
     /// The set of services that this peer requires.
     required_services: Services,
+    /// Waker used for the next poll
+    waker: Option<Waker>,
+}
+
+impl<T: Clone + Ord + Unpin> Future for ConnectionState<T> {
+    type Output = ();
+
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        self.waker.store_waker(cx);
+
+        while let Some((id, sleep)) = self.unban_timeout.as_mut() {
+            if sleep.poll_unpin(cx).is_ready() {
+                let id = id.clone();
+                self.unmark_banned(id);
+                self.unban_timeout = None;
+
+                // While we are unbanning, try to unban more peers if their deadline also have been reached.
+                // Otherwise schedule a sleep until for the first peer that can be unbanned.
+                while let Some((next_id, deadline)) = self.unban_deadlines.pop_front() {
+                    if deadline.elapsed() >= self.ban_time {
```

Comment:
```suggestion
                    if Instant::now() >= deadline {
```

`deadline` is the instant where the ban is supposed to expire. The current code checks that the expiry is at least `self.ban_time` away, i.e. that the ban has essentially expired twice.